src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs

```diff
@@ -171,34 +238,33 @@ private JObject GenerateDependencies(JObject projectJsonRoot, ITaskItem[] extern
                 // Update versions in dependencies
                 foreach (JProperty property in originalDependenciesList.Select(od => od))
                 {
-                    // Validate that the package matches the identity regex for packages we want to update.
-                    Match updateDependency = _identityRegex.Match(property.Name);
+                    ITaskItem package = GetPackageInformation(property.Name, packageInformation);
 
-                    // if the package is in the external packages list then we'll
-                    // need to replace the version with the version from the list
-                    string externalVersion = null;
-                    if (externalPackageVersions != null)
-                        externalVersion = externalPackageVersions.FirstOrDefault(epv => epv.ItemSpec.Equals(property.Name, StringComparison.OrdinalIgnoreCase))?.GetMetadata("Version");
-
-                    if (updateDependency.Success || externalVersion != null)
+                    if (package != null)
                     {
                         NuGetVersion nuGetVersion;
-                        if (externalVersion == null)
+                        if (NuGetVersion.TryParse(property.Value.ToString(), out nuGetVersion))
                         {
-                            if (NuGetVersion.TryParse(property.Value.ToString(), out nuGetVersion))
+                            NuGetVersion dependencyVersion = nuGetVersion;
+                            string ver = null;
+
+                            // a package version was provided, use its version information.
+                            if (package != null)
+                            {
+                                ver = package.GetMetadata("Version");
+                            }
+                            else
                             {
-                                Match m = _versionStructureRegex.Match(nuGetVersion.ToString());
+                                ver = string.Join(".", dependencyVersion.Major, dependencyVersion.Minor, dependencyVersion.Patch);
+                            }
 
-                                if (m.Success)
-                                {
-                                    NuGetVersion dependencyVersion = nuGetVersion;
-                                    nuGetVersion = NuGetVersion.Parse(string.Join(".", dependencyVersion.Major, dependencyVersion.Minor, dependencyVersion.Patch) + "-" + PackageBuildNumberOverride);
-                                }
+                            string releaseVersion = package.GetMetadata("ReleaseVersion");
+                            // we have package information, so use that.
+                            if (!string.IsNullOrWhiteSpace(releaseVersion))
+                            {
+                                ver = ver + "-" + releaseVersion;
```

Comment:
if we parse a version via NuGet api's, then the '-' is not part of the prerelease group.  It seems to be clearer if the prerelease group only contains the version that NuGet (or someone reading the code) would understand, and that we should add the '-' as a joining character when appending the release version.